plugins/artifacts/artifacts.go

```diff
@@ -1,143 +1,315 @@
+	err = tp.context.UploadS3Artifact(runtime.S3Artifact{
+		Name:     a.Name,
+		Mimetype: mtype,
+		Stream:   r,
+		Expires:  a.Expires,
+	})
+
+	if err != nil && err != context.Canceled {
+		tp.nonFatalErr.Set(true)
+		i := tp.monitor.ReportError(err, "Failed to upload artifact")
+		tp.context.LogError("Failed to upload artifact unhandled error, incidentId:", i)
 	}
-	return true, nil
 }
 
-func (tp taskPlugin) errorHandled(name string, expires time.Time, err error) bool {
-	var reason string
-	if _, ok := runtime.IsMalformedPayloadError(err); ok {
-		reason = "invalid-resource-on-worker"
-	} else if err == engines.ErrFeatureNotSupported || err == runtime.ErrNonFatalInternalError || err == engines.ErrHandlerInterrupt {
-		reason = "invalid-resource-on-worker"
-	} else if err == engines.ErrResourceNotFound {
-		reason = "file-missing-on-worker"
+func (tp *taskPlugin) processDirectory(result engines.ResultSet, a artifact) {
+	debug("extracting directory from path: %s", a.Path)
+	semaphore := make(chan struct{}, maxUploadConcurrency)
+	err := result.ExtractFolder(a.Path, func(p string, r ioext.ReadSeekCloser) error {
+		debug(" - Found artifact: %s in %s", p, a.Path)
+		// Always close the reader
+		defer r.Close()
+
+		// Block until we can write to semaphore, then read when we're done uploading
+		// This way the capacity o the semaphore channel limits concurrency.
+		select {
+		case semaphore <- struct{}{}:
+		case <-tp.context.Done():
+			return context.Canceled
+		}
+		defer func() {
+			<-semaphore
+		}()
+
+		// Guess the mimetype
+		mtype := mime.TypeByExtension(filepath.Ext(p))
+		if mtype == "" {
+			mtype = unknownMimetype
+		}
+
+		// Upload artifact
+		debug(" - Uploading %s from %s -> %s", p, a.Path, path.Join(a.Name, p))
+		uerr := tp.context.UploadS3Artifact(runtime.S3Artifact{
+			Name:     path.Join(a.Name, p),
+			Expires:  a.Expires,
+			Stream:   r,
+			Mimetype: mtype,
+		})
+
+		// If we have an upload error, that's just a internal non-fatal error.
+		// We ignore the error, if TaskContext was canceled, as requests should be
+		// aborted when that happens.
+		if uerr != nil && tp.context.Err() != nil {
+			tp.nonFatalErr.Set(true)
+			i := tp.monitor.ReportError(uerr, "Failed to upload artifact")
+			tp.context.LogError("Failed to upload artifact unhandled error, incidentId:", i)
+		}
+
+		// Return nil, if we're not aborted
+		return tp.context.Err()
+	})
+
+	// If feature isn't supported this is malformed-payload
+	if err == engines.ErrFeatureNotSupported {
+		e := runtime.NewMalformedPayloadError(
+			"Artifact extraction is not supported in current configuration of this workerType",
+		)
+		tp.mErrors.Lock()
+		tp.errors = append(tp.errors, e)
+		tp.mErrors.Unlock()
+		return
 	}
 
-	if reason != "" {
-		tp.context.Log("Artifact upload error handled. Continuing...", name, err.Error())
+	// If resource isn't found, task should fail and we print a message to task log
+	if err == engines.ErrResourceNotFound {
```

Comment:
Okay, I'm moving the `CreateErrorArtifact` under the `if result.Success() {`...
but I do want to handle the error...